python/lsst/meas/algorithms/stamps.py

```diff
@@ -0,0 +1,351 @@
+    use_mask : `bool`, optional
+        If ``True`` read and write the mask data.  Default ``True``.
+    use_variance : `bool`, optional
+        If ``True`` read and write the variance data. Default ``True``.
+
+    Notes
+    -----
+    A (gen2) butler can be used to read only a part of the stamps,
+    specified by a bbox:
+
+    >>> starSubregions = butler.get("brightStarStamps_sub", dataId, bbox=bbox)
+    """
+
+    def __init__(self, stamps, metadata=None, use_mask=True, use_variance=True):
+        if not hasattr(stamps, '__iter__'):
+            raise ValueError('The stamps parameter must be iterable.')
+        self._stamps = stamps
+        self._metadata = PropertySet() if metadata is None else metadata.deepCopy()
+        self.use_mask = use_mask
+        self.use_variance = use_variance
+
+    @classmethod
+    @abc.abstractmethod
+    def readFits(cls, filename):
+        """Build an instance of this class from a file.
+
+        Parameters
+        ----------
+        filename : `str`
+            Name of the file to read
+        """
+        raise NotImplementedError
+
+    @classmethod
+    @abc.abstractmethod
+    def readFitsWithOptions(cls, filename, options):
+        """Build an instance of this class with options.
+
+        Parameters
+        ----------
+        filename : `str`
+            Name of the file to read
+        options : `PropertySet`
+            Collection of metadata parameters
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def writeFits(self, filename):
+        """Write this object to a file.
+
+        Parameters
+        ----------
+        filename : `str`
+            Name of file to write
+        """
+        raise NotImplementedError
+
+    def __len__(self):
+        return len(self._stamps)
+
+    def __getitem__(self, index):
+        return self._stamps[index]
+
+    def __iter__(self):
+        return iter(self._stamps)
+
+    def append(self, item):
+        """Add an additional bright star stamp.
+
+        Parameters
+        ----------
+        item : `object`
+            Stamp-like object to append.
+        """
+        if not hasattr(item, 'stamp'):
+            raise ValueError("Ojbects added must contain a stamp attribute.")
+        self._stamps.append(item)
+        return None
+
+    def extend(self, s):
+        """Extend Stamps instance by appending elements from another instance.
+
+        Parameters
+        ----------
+        s : `list` [`object`]
+            List of Stamp-like object to append.
+        """
+        self._stamps += s._stamps
+
+    def getMaskedImages(self):
+        """Retrieve star images.
+
+        Returns
+        -------
+        maskedImages :
+            `list` [`lsst.afw.image.maskedImage.maskedImage.MaskedImageF`]
+        """
+        return [stamp.stamp_im for stamp in self._stamps]
+
+    @property
+    def metadata(self):
+        return self._metadata.deepCopy()
```

Comment:
You should think about whether the deep copy is the right thing to do. It makes it hard for people outside to change anything inside. For afw the getMetadata returns the actual thing that you can then update. If your use case is to tightly control this metadata then that's fine but maybe worth adding a docstring here.

Reply:
I don't know what the use case @MorganSchmitz was trying solve was.  I'll try to close the loop on that.

Reply:
Yep, tight control of this metadata was what I had in mind. Initially I just had a non-property getter return a `_metadata` attribute to simply _imply_ users are unlikely to want to change things in the metadata. Then following [discussions with Nate during review](https://github.com/lsst/meas_algorithms/pull/205#discussion_r462333347), I went with the copy.

My reasoning was that we are using the metadata to read and write a lot of ad hoc information - I believe for the generic class that'll just be number of stamps, and toggles for whether the variance and mask planes are present. But in my original case of bright stars, I also have Gaia object IDs, annular flux measurements, and parameters related to the latter that are absolutely necessary to use the data. So I figured I really wanted to avoid users accidentally changing that.

I don't know if that all makes sense? It is by no means crucial to have it be a copy; again, I had it return the actual thing initially. Feel free to change to that if that sounds like a better option!

Reply:
I haven't been following this closely, but if you are using FITS headers to persist structural information or actual data which is retrieved from the in-memory Python class via attributes or methods, I think it is best practice to strip that information from any flexible key/value metadata loaded from the FITS headers and re-create it on output.  There should only be one copy of that data in any given object.  See how headers are handled in Exposure.

Reply:
The current PR removes the use of this metadata where a property is more appropriate (so for example the number of "stamps" is not stored there any more).

Reply:
I'll make sure to change the bright star subclass to do the same, or at least strip the copy in the metadata like KT was saying, as soon as this is merged!

Reply:
OK, I'm going to return the actual object.  Thanks for all the input.